marathon_lb.py

```diff
@@ -166,27 +166,35 @@ def __init__(self, hosts, health_check, strict_mode, auth, ca_cert=None):
             self.__verify = ca_cert
 
     def api_req_raw(self, method, path, auth, body=None, **kwargs):
+        response, exception = None, None
         for host in self.__hosts:
             path_str = os.path.join(host, 'v2')
 
             for path_elem in path:
                 path_str = path_str + "/" + path_elem
+            try:
+                response = requests.request(
+                    method,
+                    path_str,
+                    auth=auth,
+                    headers={
+                        'Accept': 'application/json',
+                        'Content-Type': 'application/json'
+                    },
+                    timeout=(3.05, 46),
+                    **kwargs
+                )
 
-            response = requests.request(
-                method,
-                path_str,
-                auth=auth,
-                headers={
-                    'Accept': 'application/json',
-                    'Content-Type': 'application/json'
-                },
-                timeout=(3.05, 46),
-                **kwargs
-            )
+                logger.debug("%s %s", method, response.url)
+                if response.status_code == 200:
+                    break
+            except Exception as e:
+                logger.debug("%s %s", method, path_str, exc_info=True)
+                exception = e
 
-            logger.debug("%s %s", method, response.url)
-            if response.status_code == 200:
-                break
+        if not response:
+            assert exception is not None
+            raise exception
```

Comment:
I think refactoring this function in general would be better than storing the response and exception outside of the try block. Something like:

```python
class APIException(Exception):
    pass


def api_call(self, method, path, auth, body=None, **kwargs):
    for host in self.__hosts:
        try:
            response = request.request()
            if response.status_code != 200:
                continue
            response.raise_for_status()
            resp_json = cleanup_json(response.json())
            if 'message' in resp_json:
                response.reason = "%s (%s)" % (
                    response.reason,
                    resp_json['message'])

            return response

        except requests.exceptions.Timeout:
            continue

    raise APIException()
```

Reply:
The idea is that it either returns a response or raises something. Since it's only catching the timeout, anything else will raise, but in the unlikely case that all hosts timeout, then the `APIException()` is raised.